lighthouse-core/audits/seo/canonical.js

```diff
@@ -91,111 +66,158 @@ class Canonical extends Audit {
-        canonicals = Array.from(new Set(canonicals));
-
-        artifacts.Hreflang.forEach(({href}) => hreflangs.push(href));
-
-        hreflangs = hreflangs
-          .filter(href => isValidRelativeOrAbsoluteURL(href))
-          .map(href => (new URL(href, baseURL)).href); // normalize URLs
-
-        if (canonicals.length === 0) {
-          return {
-            rawValue: true,
-            notApplicable: true,
-          };
-        }
-
-        if (canonicals.length > 1) {
-          return {
-            rawValue: false,
-            explanation: str_(UIStrings.explanationConflict, {urlList: canonicals.join(', ')}),
-          };
-        }
-
-        const canonical = canonicals[0];
-
-        if (!isValidRelativeOrAbsoluteURL(canonical)) {
-          return {
-            rawValue: false,
-            explanation: str_(UIStrings.explanationInvalid, {url: canonical}),
-          };
-        }
-
-        if (!URL.isValid(canonical)) {
-          return {
-            rawValue: false,
-            explanation: str_(UIStrings.explanationRelative, {url: canonical}),
-          };
-        }
-
-        const canonicalURL = new URL(canonical);
-
-        // cross-language or cross-country canonicals are a common issue
-        if (hreflangs.includes(baseURL.href) && hreflangs.includes(canonicalURL.href) &&
-          baseURL.href !== canonicalURL.href) {
-          return {
-            rawValue: false,
-            explanation: str_(UIStrings.explanationPointsElsewhere, {url: baseURL.href}),
-          };
-        }
-
-        // bing and yahoo don't allow canonical URLs pointing to different domains, it's also
-        // a common mistake to publish a page with canonical pointing to e.g. a test domain or localhost
-        if (getPrimaryDomain(canonicalURL) !== getPrimaryDomain(baseURL)) {
-          return {
-            rawValue: false,
-            explanation: str_(UIStrings.explanationDifferentDomain, {url: canonicalURL}),
-          };
-        }
-
-        // another common mistake is to have canonical pointing from all pages of the website to its root
-        if (canonicalURL.origin === baseURL.origin &&
-          canonicalURL.pathname === '/' && baseURL.pathname !== '/') {
-          return {
-            rawValue: false,
-            explanation: str_(UIStrings.explanationRoot),
-          };
-        }
-
-        return {
-          rawValue: true,
-        };
-      });
+    const mainResource = await MainResource.request({devtoolsLog, URL: artifacts.URL}, context);
+    const baseURL = new URL(mainResource.url);
+
+    const canonicalURLData = Canonical.collectCanonicalURLs(artifacts);
+    const canonicalURLOrAuditProduct = Canonical.findValidCanonicaURLOrFinish(canonicalURLData);
+    // We didn't find a valid canonical URL, we found an error result so go ahead and return it.
+    if (typeof canonicalURLOrAuditProduct === 'object') return canonicalURLOrAuditProduct;
```

Comment:
can we separate these things like `mistakeAuditProduct()` does below?

Return `Canonical.invalidCanonicalProduct()` or whatever if it returns something, if not, the canonical url has to be `[...canonicalURLData.uniqueCanonicalURLs][0]` (or to avoid the double work, `collectCanonicalURLs` could have turned it into an array before returning)

Reply:
done